tests/turnstile_chain.c

```diff
@@ -0,0 +1,106 @@
+#include <ktest.h>
+#include <klog.h>
+#include <mutex.h>
+#include <runq.h>
+#include <sched.h>
+#include <thread.h>
+
+// length of the propagation chain
+#define T 4
+
+// TODO replace asserts about priority borrowing with some accessor
+/* TODO consider where we could check restoring priorities
+ * I think that all threads should keep HIGH priority until the last propagator
+ * has acquired 2nd mutex and then the priorities would be restored to initial
+ * ones (propagator_priority(i) for propagator[i]).
+ */
+
+static mtx_t mtx[T + 1];
+// TODO consider whether this would be any useful
+// volatile int mtx_passed[T + 1];
+
+// for simpler code: propagator[0] = starter
+static thread_t *propagator[T + 1];
+static thread_t *starter;
+
+static void set_base_prio(thread_t *td, prio_t prio) {
+  WITH_SPINLOCK(td->td_spin) {
+    sched_set_prio(td, prio);
+  }
+}
+
+// n <- [0..T]
+// TODO maybe rename to separate_prio or something?
+static int propagator_prio(int n) {
+  // TODO should this remain `n + 1` and not just `n`?
+  return (n + 1) * RQ_PPQ;
+}
+
+/* n <- [1..T]
+ * nth propagator will contest nth mutex and block on (n-1)th mutex (owned
+ * by (n-1)th propagator) causing priority propagation to [0..n-1]th propagator
+ */
+static void propagator_routine(void *argn) {
+  int n = (int)argn;
```

Comment:
I'd change this code to `propagator_routine(int n)` and change `thread_create` to `propagator[i] = thread_create(name, (void (*)(void *))propagator_routine, (void *)i);`.

Reply:
Done.